### PR TITLE
Attempt install from source when pre-built binary download fails

### LINKED ISSUE #127: Compile node from the source if binary is not pre-available
It would be great if we could compile node from the source if a binary is not pre-available.

[This blog post](https://shapeshed.com/compiling-nodejs-from-source-on-ubuntu-10-04/) looks like a good place to start experimenting.

### PR DESCRIPTION
**Pull request by LeeroyDing**
Fixes #127

### CODE REVIEW

File: functions/nvm.fish [hunk 1]
```diff
@@ -79,20 +79,28 @@ function nvm --argument-names cmd v --description "Node version manager"
...
+                    set --local dir "node-$v"
+                    set --local url $nvm_mirror/$v/$dir.$ext
+                    command mkdir -p $nvm_data/src/$v
+                    if ! command curl --progress-bar --location $url \
+                        | command tar --extract --gzip --strip-components=1 --directory $nvm_data/src/$v 2>/dev/null
+                        echo -e "Couldn't download source tarball."
+                        return 1
+                    end
+                    pushd $nvm_data/src/$v
+                    command /bin/sh $nvm_data/src/$v/configure --prefix="$nvm_data/$v"
```
Comment (LeeroyDing) [inline functions/nvm.fish]: How do I go to a different path if any of these commands failed?

**Comment by LeeroyDing** [comment]
I'm not very familiar with fish... How do I add an option for `nvm install -s $version`?

File: functions/nvm.fish [hunk 2]
```diff
@@ -79,20 +79,28 @@ function nvm --argument-names cmd v --description "Node version manager"
                 echo -e "Fetching \x1b[4m$url\x1b[24m\x1b[7m"
 
                 if ! command curl --progress-bar --location $url \
-                        | command tar --extract --gzip --directory $nvm_data/$v 2>/dev/null
-                    command rm -rf $nvm_data/$v
+                        | command tar --extract --gzip --strip-components=1 --directory $nvm_data/$v 2>/dev/null
```
Comment (jorgebucaran) [inline functions/nvm.fish]: IIRC `--strip-components=1` won't work on FreeBSD.

**Comment by jorgebucaran** [comment]
For parsing options, you usually want to use the [`argparse`](https://fishshell.com/docs/current/cmds/argparse.html) built-in.

File: functions/nvm.fish (hunk 2 above)
Comment (LeeroyDing) [reply functions/nvm.fish]: I see... It'll have to converge here then

File: functions/nvm.fish [hunk 3]
```diff
@@ -100,6 +101,46 @@ function nvm --argument-names cmd v --description "Node version manager"
...
+                    echo -e "Couldn't download source tarball."
+                    return 1
+                end
+                pushd $tmpdir
+                trap popd EXIT
+                command $tmpdir/configure --prefix="$nvm_data/$v"
+                command make -j9
+                command make install
+                popd
+                command rm -rf -- $tmpdir
```
Comment (LeeroyDing) [inline functions/nvm.fish:133]: Is it redundant when using trap?

**Comment by jorgebucaran** [comment]
Let's write a function, say `_nvm_install_from_source` that takes care of building Node from the source. I'll look into integrating it into nvm. What do you think?

**Comment by jorgebucaran** [comment]
Let's write a function, say `_nvm_build_from_source` that takes care of building Node from the source. I'll look into integrating it into nvm. What do you think?

**Comment by jorgebucaran** [comment]
Let's write a function, say `_nvm_build_from_source` that builds Node from the source. I'll look into integrating it into nvm.

**Comment by LeeroyDing** [comment]
Sounds good, let me move it to a function to use later

**Comment by LeeroyDing** [comment]
The installation is essentially 4 steps:

1. Download the source tarball and extract it to a temp folder.
2. Run `./configure --prefix="$INSTALLATION_TARGET"`, in our case `$nvm_data/$v`.
3. `make` with `-jN`, where N=number of cores + 1. Building without `-j` will work too, but is a lot slower. I haven't added the core count detection.
4. Run `make install` to install the compiled binary into the previously specified `prefix` location.

**Comment by jorgebucaran** [comment]
TBH I feel this opens up too big a can of worms, but I'm going to keep this open to see if more people chime in. Thanks!

**Comment by jorgebucaran** [comment]
I feel this opens up too big a can of worms, sorry. Thank you for working on this PR, but I won't be adding this feature.

**Pull request closed without merging**